zerver/views/streams.py

```diff
@@ -607,16 +612,41 @@ def get_streams_backend(
     return json_success({"streams": streams})
 
 @has_request_variables
-def get_topics_backend(request: HttpRequest, user_profile: UserProfile,
+def get_topics_backend(request: HttpRequest, maybe_user_profile: Union[UserProfile, AnonymousUser],
                        stream_id: int=REQ(converter=to_non_negative_int,
-                                          path_only=True)) -> HttpResponse:
-    (stream, recipient, sub) = access_stream_by_id(user_profile, stream_id)
+                                          path_only=True),
+                       narrow: Optional[List[Dict[str, Any]]]=REQ('narrow', converter=narrow_parameter, default=None),
+) -> HttpResponse:
+    if is_web_public_narrow(narrow) and not maybe_user_profile.is_authenticated:
+        is_web_public_query = True
+        user_profile: Optional[UserProfile] = None
+    else:
+        is_web_public_query = False
+        assert isinstance(maybe_user_profile, UserProfile)
+        user_profile = maybe_user_profile
+        assert user_profile is not None
+
+    if is_web_public_query:
+        realm = get_realm_from_request(request)
+        if realm is None:
+            return json_error(_("Invalid subdomain."))
```

Comment:
I think this might be able to just be `request.realm`; we started setting that in `zerver/middleware.py` recently.

Reply:
```
class HostDomainMiddleware(MiddlewareMixin):
    def process_request(self, request: HttpRequest) -> Optional[HttpResponse]:
        # Match against ALLOWED_HOSTS, which is rather permissive;
        # failure will raise DisallowedHost, which is a 400.
        request.get_host()

        # This check is important to avoid doing the extra work of
        # `get_realm` (which does a database query that could be
        # problematic for Tornado).  Also the error page below is only
        # appropriate for a page visited in a browser, not the API.
        #
        # API authentication will end up checking for an invalid
        # realm, and throw a JSON-format error if appropriate.
        if request.path.startswith(("/static/", "/api/", "/json/")):
            return None

        subdomain = get_subdomain(request)
        if subdomain != Realm.SUBDOMAIN_FOR_ROOT_DOMAIN:
            try:
                request.realm = get_realm(subdomain)
            except Realm.DoesNotExist:
                return render(request, "zerver/invalid_realm.html", status=404)
        return None
```

It seems that we don't have `request.realm` on API endpoints

Reply:
Ahh, OK, never mind then.